### PR TITLE
[FIXED JENKINS-33396] Config tabs on System Configuration

### PR DESCRIPTION
**Pull request by tfennelly**
See [JENKINS-33396](https://issues.jenkins-ci.org/browse/JENKINS-33396).

Ala ..

screenshot 2016-03-08 17 06 42

### CODE REVIEW

**Comment by daniel-beck** [comment]
One possible issue in the tab approach for the system config is that it hides broken configuration that would otherwise be plainly visible when scrolling through the form and seeing bold, red text. Not sure about this.

**Comment by daniel-beck** [comment]
Is this supposed to start out untabbed?

**Comment by tfennelly** [comment]
@daniel-beck no, it's not supposed to start out untabbed, but the user can untab it if they wish and leave it that way if they want.

**Comment by jglick** [comment]
:bee: from what I can tell. I suspect the devil will be in the details of various plugins not using `f:section` properly.

**Comment by daniel-beck** [comment]
@jglick Or core, in the case of anonymous usage statistics.

**Comment by daniel-beck** [comment]
For unknown reason, this doesn't work right. Probably also affects job configs then, but wasn't apparently in the fake jobs this was tested with?

@tfennelly [Watch me reload the system config for one-and-a-half minutes.](https://dl.dropboxusercontent.com/u/29853/PR2096.mov)

**Comment by daniel-beck** [comment]
:bug: This change also completely hides the *Home directory* informational entry at the top.

**Comment by tfennelly** [comment]
@daniel-beck what are we looking at in your video? Is that you reloading the whole time, or was the page doing it? Need more info on this please.

**Comment by tfennelly** [comment]
@daniel-beck The Home directory fixed now.

**Comment by daniel-beck** [comment]
I looked at the PR build, which includes @gusreiber's toilet paper roll in alpha 3.

* While the form resizes when changing the window width, the tabs section does not.
* Four rows of tabs look a bit silly (when installing all the plugins we offer during setup, which is just two clicks to implement).
* There are a few visual glitches when the selected tab "bounces" to the one to the left of the intended target. It's feels a bit like inertial scrolling on iOS that way, but clearly unintended.

Maybe have @gusreiber take over this change from @tfennelly? It seems that this depends on lot on details of the tabbed configs implementation.

**Comment by tfennelly** [comment]
@daniel-beck those things have nothing to with this PR from what I can make out, right? They are just issues/concerns with config page tabbing in general, right? If so, then why hold up this PR?

**Comment by tfennelly** [comment]
I had a quick look at this and it's a general issue with the new config page layout i.e. not an issue specific to this PR.

**Comment by daniel-beck** [comment]
@tfennelly 

> those things have nothing to with this PR from what I can make out, right? They are just issues/concerns with config page tabbing/scrolling in general, right? If so, then why hold up this PR?

Not exactly.

> While the form resizes when changing the window width, the tabs section does not.

Gus made job config forms fixed width, which was strongly critised in #2106 but still, it's not an issue there. FWIW I can not reproduce the problem currently, so it may be obsolete from an older build.

----

> Four rows of tabs look a bit silly (when installing all the plugins we offer during setup, which is just two clicks to implement).

Job config forms don't have this many tabs, plugins typically don't contribute new ones, and the titles aren't as ridiculously long. So it's not a problem there.

----

> There are a few visual glitches when the selected tab "bounces" to the one to the left of the intended target.

I have not observed this problem with job configuration tabs, possibly because they're shorter.

----

So, yes, while all of this stems from Gus' PR, it only becomes relevant in this one. I'd probably have asked Gus to fix things if the original scroll tabs PR included this change, but it did not.

**Comment by kohsuke** [comment]
Daniel told me that this fix is being taken care of as a part of the scrollspy fix. Tom, can you verify that? Does that mean we think the follow-up change in scrollspy might fix this issue and we'll come back here later?

**Comment by tfennelly** [comment]
@kohsuke The cleanup and fixes in #2121 will fix the resizing and bouncing issues. The "4 rows of tabs" issue is not fixed here as that is an enhancement to the tabbar and would be done in a separate PR.

**Comment by daniel-beck** [comment]
@tfennelly To clarify, can we merge this now and it will just magically work with #2121?

**Comment by daniel-beck** [comment]
@kzantow @gusreiber Maybe you can answer my previous comment?

> can we merge this now and it will just magically work with #2121?

**Comment by kzantow** [comment]
@daniel-beck I do not think we should get this merged. The system config has too many sections to make this a good UI. To answer your question, AFAIK this should 'just work' with #2121 based on what @tfennelly said

**Comment by daniel-beck** [comment]
@kzantow Note that several tabs have been moved into a different dialog. And with some plugins (looking at you, P4!) adapted in the future to use shorter tab titles, it shouldn't go beyond two rows most of the time.

Still, this appears to be an improvement over the current situation. If user feedback is negative, we can always back it out. I'd also be happy with a tab-less, styled config form, but AFAIU this isn't currently possible.

**Comment by tfennelly** [comment]
@daniel-beck I've update the PR now.

One thing I notice that's different with the Job config pages is that the form does not resize below a certain width, which makes the tabs look a bit strange. Not sure if we can do anything about that.

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly [hunk 1]
```diff
@@ -61,8 +64,8 @@ THE SOFTWARE.
       </j:forEach>
 
       <f:bottomButtonBar>
-        <f:submit value="${%Save}" />
-        <f:apply />
+        <f:submit value="${%Save All}"/>
+        <f:apply value="${%Apply All}"/>
```
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: If we have scroll tabs, this is unnecessary.

Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: Can be addressed during/after merge.

Comment (tfennelly) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: You saying we should revert this change? That's easy to do if you want.

Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: @tfennelly Yes. With everything on one page, no need to change label to clarify all tabs are getting saved.

**Comment by daniel-beck** [comment]
This is not going into 2.0, and for 2.1+ it needs to be re-filed against the master branch.

**Comment by tfennelly** [comment]
Closing as requested by @daniel-beck
